# 1. 선형 회귀 구현하기

In [1]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

In [2]:
torch.manual_seed(1)

실습을 위한 기본적인 셋팅이 끝났습니다. 이제 훈련 데이터인 x_train과 y_train을 선언합니다.



In [3]:
x_train = torch.FloatTensor([[1], [2], [3]])
y_train = torch.FloatTensor([[2], [4], [6]])

x_train과 x_train의 크기(shape)를 출력해보겠습니다.



In [4]:
print(x_train)
print(x_train.shape)

tensor([[1.],
        [2.],
        [3.]])
torch.Size([3, 1])


x_train의 값이 출력되고, x_train의 크기가 (3 × 1)임을 알 수 있습니다.
y_train과 y_train의 크기(shape)를 출력해보겠습니다.

In [5]:
print(y_train)
print(y_train.shape)

tensor([[2.],
        [4.],
        [6.]])
torch.Size([3, 1])


y_train의 값이 출력되고, y_train의 크기가 (3 × 1)임을 알 수 있습니다.



선형 회귀란 학습 데이터와 가장 잘 맞는 하나의 직선을 찾는 일입니다.
그리고 가장 잘 맞는 직선을 정의하는 것은 바로 W와 b입니다.
결론적으로 선형 회귀의 목표는 가장 잘 맞는 직선을 정의하는 W와 b의 값을 찾는 것입니다.
우선 가중치 W를 0으로 초기화하고, 이 값을 출력해보겠습니다.

In [6]:
W = torch.zeros(1, requires_grad=True)
print(W)

tensor([0.], requires_grad=True)


가중치 W가 0으로 초기화되어있으므로 0이 출력된 것을 확인할 수 있습니다. 위에서 requires_grad=True가 인자로 주어진 것을 확인할 수 있습니다. 이는 이 변수는 학습을 통해 계속 값이 변경되는 변수임을 의미합니다.

마찬가지로 편향
b도 0으로 초기화하고, 학습을 통해 값이 변경되는 변수임을 명시합니다.

In [7]:
b = torch.zeros(1, requires_grad=True)
print(b)

tensor([0.], requires_grad=True)


현재 가중치 W와 b둘 다 0이므로 현 직선의 방정식은 다음과 같습니다.
$y = 0 × x + 0$
지금 상태에선 x에 어떤 값이 들어가도 가설은 0을 예측하게 됩니다. 즉, 아직 적절한 W와 b의 값이 아닙니다.

파이토치 코드 상으로 직선의 방정식에 해당되는 가설을 선언합니다.

$ H(x)= Wx+b$  

In [8]:
hypothesis = x_train * W + b

파이토치 코드 상으로 선형 회귀의 비용 함수에 해당되는 평균 제곱 오차를 선언합니다.


$$
cost(W, b) = \frac{1}{n} \sum_{i=1}^{n} \left[y^{(i)} - H(x^{(i)})\right]^2
$$

In [9]:
# 앞서 배운 torch.mean으로 평균을 구한다.
cost = torch.mean((hypothesis - y_train) ** 2)
print(cost)

tensor(18.6667, grad_fn=<MeanBackward0>)


이제 경사 하강법을 구현합니다. 아래의 'SGD'는 경사 하강법의 일종입니다. lr은 학습률(learning rate)를 의미합니다.
학습 대상인 W와 b가 SGD의 입력이 됩니다.

In [10]:
optimizer = optim.SGD([W, b], lr=0.01)

optimizer.zero_grad()를 실행하므로서 미분을 통해 얻은 기울기를 0으로 초기화합니다. 기울기를 초기화해야만 새로운 가중치 편향에 대해서 새로운 기울기를 구할 수 있습니다.  

그 다음 cost.backward() 함수를 호출하면 가중치 W와 편향 b에 대한 기울기가 계산됩니다.  

그 다음 경사 하강법 최적화 함수 opimizer의 .step() 함수를 호출하여 인수로 들어갔던 W와 b에서 리턴되는 변수들의 기울기에 학습률(learining rate) 0.01을 곱하여 빼줌으로서 업데이트합니다.

In [11]:
# gradient를 0으로 초기화
optimizer.zero_grad()

# 비용 함수를 미분하여 gradient 계산
cost.backward()

# W와 b를 업데이트
optimizer.step()

결과적으로 훈련 과정에서 $W$와 $b$는 훈련 데이터와 잘 맞는 직선을 표현하기 위한 적절한 값으로 변화해갑니다.

In [12]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

torch.manual_seed(1)

# 데이터
x_train = torch.FloatTensor([[1], [2], [3]])
y_train = torch.FloatTensor([[2], [4], [6]])

# 모델 초기화
W = torch.zeros(1, requires_grad=True)
b = torch.zeros(1, requires_grad=True)

# optimizer 설정
optimizer = optim.SGD([W, b], lr=0.01)

nb_epochs = 2000 # 원하는만큼 경사 하강법을 반복
for epoch in range(1, nb_epochs + 1):

    # H(x) 계산
    hypothesis = x_train * W + b

    # cost 계산
    cost = torch.mean((hypothesis - y_train) ** 2)

    # cost로 H(x) 개선
    optimizer.zero_grad()
    cost.backward()
    optimizer.step()

    # 100번마다 로그 출력
    if epoch % 100 == 0:
        print('Epoch {:4d}/{} W: {:.3f}, b: {:.3f} Cost: {:.6f}'.format(
            epoch, nb_epochs, W.item(), b.item(), cost.item()
        ))

Epoch  100/2000 W: 1.745, b: 0.579 Cost: 0.048403
Epoch  200/2000 W: 1.800, b: 0.456 Cost: 0.029910
Epoch  300/2000 W: 1.842, b: 0.358 Cost: 0.018483
Epoch  400/2000 W: 1.876, b: 0.281 Cost: 0.011421
Epoch  500/2000 W: 1.903, b: 0.221 Cost: 0.007058
Epoch  600/2000 W: 1.923, b: 0.174 Cost: 0.004361
Epoch  700/2000 W: 1.940, b: 0.137 Cost: 0.002695
Epoch  800/2000 W: 1.953, b: 0.107 Cost: 0.001665
Epoch  900/2000 W: 1.963, b: 0.084 Cost: 0.001029
Epoch 1000/2000 W: 1.971, b: 0.066 Cost: 0.000636
Epoch 1100/2000 W: 1.977, b: 0.052 Cost: 0.000393
Epoch 1200/2000 W: 1.982, b: 0.041 Cost: 0.000243
Epoch 1300/2000 W: 1.986, b: 0.032 Cost: 0.000150
Epoch 1400/2000 W: 1.989, b: 0.025 Cost: 0.000093
Epoch 1500/2000 W: 1.991, b: 0.020 Cost: 0.000057
Epoch 1600/2000 W: 1.993, b: 0.016 Cost: 0.000035
Epoch 1700/2000 W: 1.995, b: 0.012 Cost: 0.000022
Epoch 1800/2000 W: 1.996, b: 0.010 Cost: 0.000014
Epoch 1900/2000 W: 1.997, b: 0.008 Cost: 0.000008
Epoch 2000/2000 W: 1.997, b: 0.006 Cost: 0.000005


에포크(Epoch)는 전체 훈련 데이터가 학습에 한 번 사용된 주기를 말합니다.
이번 실습의 경우 2,000번을 수행했습니다.

최종 훈련 결과를 보면 최적의 기울기 W는 2에 가깝고, b는 0에 가까운 것을 볼 수 있습니다. 현재 훈련 데이터가 x_train은 [[1], [2], [3]]이고 y_train은 [[2], [4], [6]]인 것을 감안하면 실제 정답은 W가 2이고, b가 0이므로 거의 정답을 찾은 셈입니다.

# 2. torch.manual_seed()를 하는 이유

torch.manual_seed()를 사용한 프로그램의 결과는 다른 컴퓨터에서 실행시켜도 동일한 결과를 얻을 수 있습니다.  

그 이유는 torch.manual_seed()는 난수 발생 순서와 값을 동일하게 보장해준다는 특징때문입니다. 우선 랜덤 시드가 3일 때 두 번 난수를 발생시켜보고, 다른 랜덤 시드를 사용한 후에 다시 랜덤 시드를 3을 사용한다면 난수 발생값이 동일하게 나오는지 보겠습니다.

In [13]:
import torch

In [14]:
torch.manual_seed(3)
print('랜덤 시드가 3일 때')
for i in range(1,3):
  print(torch.rand(1))

랜덤 시드가 3일 때
tensor([0.0043])
tensor([0.1056])


랜덤 시드가 3일때 두 개의 난수를 발생시켰더니 0.0043과 0.1056이 나옵니다. 이제 랜덤 시드값을 바꿔봅시다.



In [15]:
torch.manual_seed(5)
print('랜덤 시드가 5일 때')
for i in range(1,3):
  print(torch.rand(1))

랜덤 시드가 5일 때
tensor([0.8303])
tensor([0.1261])


0.8303과 0.1261이 나옵니다. 이제 다시 랜덤 시드값을 3으로 돌려보겠습니다. 이렇게 하면 프로그램을 다시 처음부터 실행한 것처럼 난수 발생 순서가 초기화됩니다.

In [16]:
torch.manual_seed(3)
print('랜덤 시드가 3일 때')
for i in range(1,3):
  print(torch.rand(1))

랜덤 시드가 3일 때
tensor([0.0043])
tensor([0.1056])


다시 동일하게 0.0043과 0.1056이 나옵니다.



# 3. optimizer.zero_grad()가 필요한 이유

파이토치는 미분을 통해 얻은 기울기를 이전에 계산된 기울기 값에 누적시키는 특징이 있습니다. 예를 들어봅시다.

In [17]:
import torch

w = torch.tensor(2.0, requires_grad=True)

nb_epochs = 20
for epoch in range(1, nb_epochs + 1):

  z = 2 * w

  z.backward()
  print('수식을 w로 미분한 값 : {}'.format(w.grad))

수식을 w로 미분한 값 : 2.0
수식을 w로 미분한 값 : 4.0
수식을 w로 미분한 값 : 6.0
수식을 w로 미분한 값 : 8.0
수식을 w로 미분한 값 : 10.0
수식을 w로 미분한 값 : 12.0
수식을 w로 미분한 값 : 14.0
수식을 w로 미분한 값 : 16.0
수식을 w로 미분한 값 : 18.0
수식을 w로 미분한 값 : 20.0
수식을 w로 미분한 값 : 22.0
수식을 w로 미분한 값 : 24.0
수식을 w로 미분한 값 : 26.0
수식을 w로 미분한 값 : 28.0
수식을 w로 미분한 값 : 30.0
수식을 w로 미분한 값 : 32.0
수식을 w로 미분한 값 : 34.0
수식을 w로 미분한 값 : 36.0
수식을 w로 미분한 값 : 38.0
수식을 w로 미분한 값 : 40.0


계속해서 미분값인 2가 누적되는 것을 볼 수 있습니다. 그렇기 때문에 optimizer.zero_grad()를 통해 미분값을 계속 0으로 초기화시켜줘야 합니다.

# 4. 로지스틱 회귀 구현하기

In [18]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

In [19]:
torch.manual_seed(1)

x_train과 y_train을 텐서로 선언합니다.



In [20]:
x_data = [[1, 2], [2, 3], [3, 1], [4, 3], [5, 3], [6, 2]]
y_data = [[0], [0], [0], [1], [1], [1]]
x_train = torch.FloatTensor(x_data)
y_train = torch.FloatTensor(y_data)

앞서 훈련 데이터를 행렬로 선언하고, 행렬 연산으로 가설을 세우는 방법을 배웠습니다.
여기서도 마찬가지로 행렬 연산을 사용하여 가설식을 세울겁니다. x_train과 y_train의 크기를 확인해봅시다.

In [21]:
print(x_train.shape)
print(y_train.shape)

torch.Size([6, 2])
torch.Size([6, 1])


현재 x_train은 6 × 2의 크기(shape)를 가지는 행렬이며, y_train은 6 × 1의 크기를 가지는 벡터입니다. x_train을 $X$라고 하고, 이와 곱해지는 가중치 벡터를 $W$라고 하였을 때, $XW$가 성립되기 위해서는 $W$ 벡터의 크기는 2 × 1이어야 합니다. 이제 W와 b를 선언합니다.

In [ ]:
W = torch.zeros((2, 1), requires_grad=True) # 크기는 2 x 1
b = torch.zeros(1, requires_grad=True)

이제 가설식을 세워보겠습니다. 파이토치에서는 $e^{x}$를 구현하기 위해서 torch.exp(x)를 사용합니다.  
이에 따라 행렬 연산을 사용한 가설식은 다음과 같습니다.

In [ ]:
hypothesis = 1 / (1 + torch.exp(-(x_train.matmul(W) + b)))

앞서 W와 b는 torch.zeros를 통해 전부 0으로 초기화 된 상태입니다. 이 상태에서 예측값을 출력해봅시다.

In [ ]:
print(hypothesis) # 예측값인 H(x) 출력

tensor([[0.5000],
        [0.5000],
        [0.5000],
        [0.5000],
        [0.5000],
        [0.5000]], grad_fn=<MulBackward0>)


실제값 y_train과 크기가 동일한 6 × 1의 크기를 가지는 예측값 벡터가 나오는데 모든 값이 0.5입니다.  

사실 가설식을 좀 더 간단하게도 구현할 수 있습니다. 이미 파이토치에서는 시그모이드 함수를 이미 구현하여 제공하고 있기 때문입니다. 다음은 torch.sigmoid를 사용하여 좀 더 간단히 구현한 가설식입니다.

In [ ]:
hypothesis = torch.sigmoid(x_train.matmul(W) + b)

앞서 구현한 식과 본질적으로 동일한 식입니다. 마찬가지로 W와 b가 0으로 초기화 된 상태에서 예측값을 출력해봅시다.

In [ ]:
print(hypothesis)

tensor([[0.5000],
        [0.5000],
        [0.5000],
        [0.5000],
        [0.5000],
        [0.5000]], grad_fn=<SigmoidBackward0>)


앞선 결과와 동일하게 y_train과 크기가 동일한 6 × 1의 크기를 가지는 예측값 벡터가 나오는데 모든 값이 0.5입니다.  

이제 아래의 비용 함수값. 즉, 현재 예측값과 실제값 사이의 cost를 구해보겠습니다.
$$
cost(W) = -\frac{1}{n} \sum_{i=1}^{n} [y^{(i)}logH(x^{(i)}) + (1-y^{(i)})log(1-H(x^{(i)}))]
$$
우선, 현재 예측값과 실제값을 출력해보겠습니다.

In [ ]:
print(hypothesis)
print(y_train)

tensor([[0.5000],
        [0.5000],
        [0.5000],
        [0.5000],
        [0.5000],
        [0.5000]], grad_fn=<SigmoidBackward0>)
tensor([[0.],
        [0.],
        [0.],
        [1.],
        [1.],
        [1.]])


현재 총 6개의 원소가 존재하지만 하나의 샘플. 즉, 하나의 원소에 대해서만 오차를 구하는 식을 작성해보겠습니다.

In [ ]:
-(y_train[0] *torch.log(hypothesis[0]) +
  (1-y_train[0]) *torch.log(1-hypothesis[0]))

tensor([0.6931], grad_fn=<NegBackward0>)

이제 모든 원소에 대해서 오차를 구해보겠습니다.

In [ ]:
losses = -(y_train * torch.log(hypothesis) +
           (1 - y_train) * torch.log(1 - hypothesis))
print(losses)

tensor([[0.6931],
        [0.6931],
        [0.6931],
        [0.6931],
        [0.6931],
        [0.6931]], grad_fn=<NegBackward0>)


그리고 이 전체 오차에 대한 평균을 구합니다.

In [ ]:
cost = losses.mean()
print(cost)

tensor(0.6931, grad_fn=<MeanBackward0>)


결과적으로 얻은 cost는 0.6931입니다.  

지금까지 비용 함수의 값을 직접 구현하였는데, 사실 파이토치에서는 로지스틱 회귀의 비용 함수를 이미 구현해서 제공하고 있습니다.  
사용 방법은 torch.nn.functional as F와 같이 임포트 한 후에 F.binary_cross_entropy(예측값, 실제값)과 같이 사용하면 됩니다.

In [ ]:
F.binary_cross_entropy(hypothesis, y_train)

tensor(0.6931, grad_fn=<BinaryCrossEntropyBackward0>)

동일하게 cost가 0.6931이 출력되는 것을 볼 수 있습니다. 모델의 훈련 과정까지 추가한 전체 코드는 아래와 같습니다.

In [ ]:
x_data = [[1, 2], [2, 3], [3, 1], [4, 3], [5, 3], [6, 2]]
y_data = [[0], [0], [0], [1], [1], [1]]
x_train = torch.FloatTensor(x_data)
y_train = torch.FloatTensor(y_data)

In [ ]:
# 모델 초기화
W = torch.zeros((2, 1), requires_grad=True)
b = torch.zeros(1, requires_grad=True)
# optimizer 설정
optimizer = optim.SGD([W, b], lr=1)

nb_epochs = 1000
for epoch in range(1, nb_epochs + 1):
    # Cost 계산
    hypothesis = torch.sigmoid(x_train.matmul(W) + b)
    cost = -(y_train * torch.log(hypothesis) +
             (1 - y_train) * torch.log(1 - hypothesis)).mean()

    # cost로 H(x) 개선
    optimizer.zero_grad()
    cost.backward()
    optimizer.step()

    # 100번마다 로그 출력
    if epoch % 100 == 0:
      print('Epoch {:4d}/{} Cost: {:.6f}'.format(
            epoch, nb_epochs, cost.item()
        ))

Epoch  100/1000 Cost: 0.135644
Epoch  200/1000 Cost: 0.080964
Epoch  300/1000 Cost: 0.058062
Epoch  400/1000 Cost: 0.045398
Epoch  500/1000 Cost: 0.037327
Epoch  600/1000 Cost: 0.031720
Epoch  700/1000 Cost: 0.027592
Epoch  800/1000 Cost: 0.024422
Epoch  900/1000 Cost: 0.021911
Epoch 1000/1000 Cost: 0.019871


학습이 끝났습니다. 이제 훈련했던 훈련 데이터를 그대로 입력으로 사용했을 때, 제대로 예측하는지 확인해보겠습니다.  
현재 W와 b는 훈련 후의 값을 가지고 있습니다. 현재 W와 b를 가지고 예측값을 출력해보겠습니다.

In [ ]:
hypothesis = torch.sigmoid(x_train.matmul(W) + b)
print(hypothesis)

tensor([[2.7711e-04],
        [3.1636e-02],
        [3.9014e-02],
        [9.5618e-01],
        [9.9823e-01],
        [9.9969e-01]], grad_fn=<SigmoidBackward0>)


현재 위 값들은 0과 1 사이의 값을 가지고 있습니다. 이제 0.5를 넘으면 True, 넘지 않으면 False로 값을 정하여 출력해보겠습니다.

In [ ]:
prediction = hypothesis >= torch.FloatTensor([0.5])
print(prediction)

tensor([[False],
        [False],
        [False],
        [ True],
        [ True],
        [ True]])


실제값은 [[0], [0], [0], [1], [1], [1]]이므로, 이는 결과적으로 False, False, False, True, True, True와 동일합니다. 즉, 기존의 실제값과 동일하게 예측한 것을 볼 수 있습니다. 훈련이 된 후의 W와 b의 값을 출력해보겠습니다.

In [ ]:
print(W)
print(b)

tensor([[3.2521],
        [1.5174]], requires_grad=True)
tensor([-14.4777], requires_grad=True)


# 5. `nn.Linear` <br>

- `nn.Linear`는 PyTorch에서 __선형 변환(Linear Transformation)__ 을 구현하는 모듈로, 선형 회귀의 핵심인 \( y = Wx + b \)를 표현할 수 있습니다.  
하지만, `nn.Linear`는 단순히 선형 변환을 수행하는 도구일 뿐, **손실 함수**나 **최적화 알고리즘**과 결합되어야 선형 회귀 모델로 완성됩니다.<br>
여러가지 이름으로 사용되기는 하지만 fc layer(Fully Connected layer) 라고 표현합니다.

---

### [nn.Linear](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html)의 구성 요소

`nn.Linear`는 다음 두 가지 학습 가능한 매개변수를 포함합니다:
1. **가중치 \( W \)**: 입력과 출력 간의 관계를 나타내는 행렬.
2. **편향 \( b \)**: 출력값에 추가되는 상수.

이 두 매개변수는 **경사하강법(Gradient Descent)**을 통해 학습됩니다.
---

### `nn.Linear`와 선형 회귀의 차이점

- **선형 회귀**:
  - \( y = Wx + b \)를 사용하여 입력 \( x \)와 출력 \( y \) 사이의 선형 관계를 학습합니다.
  - 손실 함수로 주로 MSE(평균 제곱 오차)를 사용합니다.
  - 학습 알고리즘(예: 경사하강법)을 통해 최적화됩니다.

- **`nn.Linear`**:
  - 입력과 출력 간의 선형 변환을 계산하는 모듈입니다.
  - 학습 과정에서 손실 함수와 옵티마이저를 따로 정의해야 합니다.
  - 선형 회귀뿐만 아니라 다른 모델(예: 신경망)에서도 사용됩니다.


In [77]:
import torch 
import torch.nn as nn 

torch.manual_seed(1)


x_train = torch.FloatTensor([[1], [2], [3]]) 

print(f"x shape : {x_train.shape}") # 1차원 짜리 Vector가 3개가 있고 1개의 Feature가진다
print()
fc = nn.Linear(1, 1) # input feature, output feature 


print(f"FC layer : {fc}")
print()

x shape : torch.Size([3, 1])

FC layer : Linear(in_features=1, out_features=1, bias=True)



In [79]:
print(f"Weight : {fc.weight}")
print(f"Bias : {fc.bias}")
print()


Weight : Parameter containing:
tensor([[0.5153]], requires_grad=True)
Bias : Parameter containing:
tensor([-0.4414], requires_grad=True)



In [80]:
# fc layer의 weight와 bias를 초기화 할 수 있습니다.  
fc.weight.data.fill_(0)
fc.bias.data.fill_(0)

print(f"Weight : {fc.weight}")
print(f"Bias : {fc.bias}")

Weight : Parameter containing:
tensor([[0.]], requires_grad=True)
Bias : Parameter containing:
tensor([0.], requires_grad=True)


In [83]:
h = fc(x_train)
print(h) # grad_fn=<AddmmBackward0> autograd 연산 속성 

tensor([[0.],
        [0.],
        [0.]], grad_fn=<AddmmBackward0>)


In [85]:
import torch
import torch.nn as nn
import torch.optim as optim

# 데이터 준비
x_train = torch.FloatTensor([[1], [2], [3]])
y_train = torch.FloatTensor([[2], [4], [6]])

# 1. 직접 구현한 선형 회귀
# 가중치(W)와 편향(b) 초기화
W = torch.zeros(1, requires_grad=True)
b = torch.zeros(1, requires_grad=True)

# 학습 설정
optimizer = optim.SGD([W, b], lr=0.01)
epochs = 1000

# 학습 루프
print("직접 구현한 선형 회귀:")
for epoch in range(1, epochs + 1):
    # 예측값 계산
    hypothesis = x_train * W + b
    
    # 손실 계산 (MSE)
    cost = torch.mean((hypothesis - y_train) ** 2)
    
    # 경사 초기화, 역전파, 매개변수 업데이트
    optimizer.zero_grad()
    cost.backward()
    optimizer.step()
    
    # 로그 출력
    if epoch % 200 == 0:
        print(f"Epoch {epoch:4d}/{epochs} W: {W.item():.3f}, b: {b.item():.3f}, Cost: {cost.item():.6f}")

# 2. nn.Linear를 사용한 선형 회귀
# 모델 정의
fc = nn.Linear(1, 1)  # 입력 크기: 1, 출력 크기: 1
optimizer = optim.SGD(fc.parameters(), lr=0.01)
fc.weight.data.fill_(0)
fc.bias.data.fill_(0)
# 학습 루프
print("\n`nn.Linear`를 사용한 선형 회귀:")
for epoch in range(1, epochs + 1):
    # 예측값 계산
    prediction = fc(x_train)
    
    # 손실 계산 (MSE)
    cost = nn.MSELoss()(prediction, y_train)
    
    # 경사 초기화, 역전파, 매개변수 업데이트
    optimizer.zero_grad()
    cost.backward()
    optimizer.step()
    
    # 로그 출력
    if epoch % 200 == 0:
        W, b = fc.parameters()
        print(f"Epoch {epoch:4d}/{epochs} W: {W.item():.3f}, b: {b.item():.3f}, Cost: {cost.item():.6f}")


직접 구현한 선형 회귀:
Epoch  200/1000 W: 1.800, b: 0.456, Cost: 0.029910
Epoch  400/1000 W: 1.876, b: 0.281, Cost: 0.011421
Epoch  600/1000 W: 1.923, b: 0.174, Cost: 0.004361
Epoch  800/1000 W: 1.953, b: 0.107, Cost: 0.001665
Epoch 1000/1000 W: 1.971, b: 0.066, Cost: 0.000636

`nn.Linear`를 사용한 선형 회귀:
Epoch  200/1000 W: 1.800, b: 0.456, Cost: 0.029910
Epoch  400/1000 W: 1.876, b: 0.281, Cost: 0.011421
Epoch  600/1000 W: 1.923, b: 0.174, Cost: 0.004361
Epoch  800/1000 W: 1.953, b: 0.107, Cost: 0.001665
Epoch 1000/1000 W: 1.971, b: 0.066, Cost: 0.000636


# 6. 소프트맥스 함수 직접 구현

---

## 1. 소프트맥스 함수란?

소프트맥스(Softmax) 함수는 **다중 클래스 분류** 문제에서 사용되며, 모델이 각 클래스에 대해 예측한 점수(logits)를 **확률 분포**로 변환합니다.  
입력 값의 상대적 크기를 기반으로 모든 클래스의 확률 합이 \(1\)이 되도록 정규화합니다.

### 소프트맥스 함수 공식:

## $$ \text{Softmax}(z_i) = \frac{e^{z_i}}{\sum_{j=1}^n e^{z_j}} $$

- $(z_i)$: 클래스 $(i)$의 점수(logit).
- $(e^{z_i})$: 클래스 $(i)$의 점수를 지수 함수로 변환.
- $(\sum_{j=1}^n e^{z_j})$: 모든 클래스의 지수 값의 합.

---

## 2. 소프트맥스 함수의 특징

1. **출력 값의 범위**:
   $
   0 \leq \text{Softmax}(z_i) \leq 1
   $
   - 모든 출력 값은 확률로 해석 가능.

2. **확률 분포**:
   $
   \sum_{i=1}^n \text{Softmax}(z_i) = 1
   $
   - 출력 값의 합이 항상 1이 됩니다.

3. **다중 클래스 분류**:
   - 소프트맥스는 각 클래스에 대한 **확률값**을 제공하여 가장 가능성이 높은 클래스를 예측할 수 있도록 합니다.

---

## 3. 소프트맥스 함수 직접 구현

소프트맥스 함수는 두 단계로 구현됩니다:
1. 각 점수(logit)를 지수 함수로 변환: $(e^{z_i})$.
2. 모든 지수 값의 합으로 각 값을 나누어 정규화.

---


In [97]:
import torch
import torch.nn as nn
import torch.optim as optim

# 1. 데이터 준비
x_train = torch.FloatTensor([[1, 2], [2, 3], [3, 1], [4, 3], [5, 3], [6, 2]])
y_train = torch.LongTensor([0, 1, 2, 0, 1, 2])  # 클래스 레이블 (0, 1, 2)

# 2. 모델 정의
class SoftmaxClassifier(nn.Module):
    def __init__(self):
        super(SoftmaxClassifier, self).__init__()
        self.fc = nn.Linear(2, 3)  # 입력 크기 2, 출력 크기 3 (클래스 개수)
    
    def forward(self, x):
        return self.fc(x)  # 소프트맥스는 CrossEntropyLoss에서 자동 계산

model = SoftmaxClassifier()

# 3. 손실 함수와 옵티마이저 정의
# PyTorch 내장 크로스 엔트로피
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=0.1)

# 4. 직접 구현한 크로스 엔트로피 손실 함수
def cross_entropy_loss(z, y):
    # 1. 소프트맥스 계산
    softmax = torch.exp(z) / torch.sum(torch.exp(z), dim=1, keepdim=True)
    # 2. 정답 클래스의 확률 추출
    log_probs = torch.log(softmax)
    target_log_probs = log_probs[range(len(y)), y]
    # 3. 손실 계산 (음수 로그 확률 평균)
    cost = -torch.mean(target_log_probs)
    return cost

# 5. 훈련 과정
epochs = 1000
for epoch in range(epochs + 1):
    # Forward
    z = model(x_train)  # 모델의 예측값 (logits)
    
    # PyTorch 내장 손실
    cost_builtin = criterion(z, y_train)
    
    # 직접 구현한 손실
    cost_manual = cross_entropy_loss(z, y_train)
    
    # Backward (직접 구현한 손실로 학습 진행)
    optimizer.zero_grad()
    cost_manual.backward()
    optimizer.step()

    # 100번마다 로그 출력
    if epoch % 100 == 0:
        print(f'Epoch {epoch:4d}/{epochs} Cost (Builtin): {cost_builtin.item():.6f} Cost (Manual): {cost_manual.item():.6f}')

# 6. 모델 평가 (직접 구현한 소프트맥스 사용)
with torch.no_grad():
    x_test = torch.FloatTensor([[4, 2]])  # 예측을 위한 새로운 데이터
    z = model(x_test)
    
    # 소프트맥스 직접 구현
    exp_z = torch.exp(z)  # 각 logit에 대해 e^z 계산
    softmax_output = exp_z / torch.sum(exp_z, dim=1, keepdim=True)  # 확률 계산

    y_pred = torch.argmax(softmax_output, dim=1)  # 가장 높은 확률의 클래스 선택

    print("\nTest Data:", x_test)
    print("Logits (z):", z)
    print("Softmax Output (Manual):", softmax_output)
    print("Predictions (y_pred):", y_pred)


Epoch    0/1000 Cost (Builtin): 1.333151 Cost (Manual): 1.333150
Epoch  100/1000 Cost (Builtin): 0.579484 Cost (Manual): 0.579484
Epoch  200/1000 Cost (Builtin): 0.518470 Cost (Manual): 0.518471
Epoch  300/1000 Cost (Builtin): 0.487488 Cost (Manual): 0.487488
Epoch  400/1000 Cost (Builtin): 0.467566 Cost (Manual): 0.467566
Epoch  500/1000 Cost (Builtin): 0.453135 Cost (Manual): 0.453135
Epoch  600/1000 Cost (Builtin): 0.441908 Cost (Manual): 0.441908
Epoch  700/1000 Cost (Builtin): 0.432757 Cost (Manual): 0.432757
Epoch  800/1000 Cost (Builtin): 0.425052 Cost (Manual): 0.425052
Epoch  900/1000 Cost (Builtin): 0.418410 Cost (Manual): 0.418410
Epoch 1000/1000 Cost (Builtin): 0.412582 Cost (Manual): 0.412582

Test Data: tensor([[4., 2.]])
Logits (z): tensor([[0.8664, 0.4130, 1.2396]])
Softmax Output (Manual): tensor([[0.3239, 0.2058, 0.4703]])
Predictions (y_pred): tensor([2])


In [98]:
import torch 

# 5. 모델 평가
with torch.no_grad():
    test_data = torch.FloatTensor([[4, 2]])  # 예측을 위한 새로운 데이터
    logits = model(test_data)
    softmax_output = torch.softmax(logits, dim=1)  # 소프트맥스 확률
    predictions = torch.argmax(softmax_output, dim=1)  # 가장 높은 확률의 클래스 선택

    print("\nTest Data:", test_data)
    print("Logits:", logits)
    print("Softmax Output:", softmax_output)
    print("Predictions:", predictions)


Test Data: tensor([[4., 2.]])
Logits: tensor([[0.8664, 0.4130, 1.2396]])
Softmax Output: tensor([[0.3239, 0.2058, 0.4703]])
Predictions: tensor([2])
